controllers/product.go

```diff
@@ -0,0 +1,174 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/distributed-marketplace-system/db"
+	"github.com/distributed-marketplace-system/models"
+	"github.com/distributed-marketplace-system/util"
+
+	"github.com/cloudinary/cloudinary-go/api/uploader"
+	"github.com/gin-gonic/gin"
+)
+
+//ProductController ...
+type ProductController struct{}
+
+func UploadProductImage(img_path string) (url string, err error) {
+
+	if _, err := os.Stat(img_path); os.IsNotExist(err) {
+		return "", err
+	}
+
+	var ctx = context.Background()
+	uploadResult, err := util.CLD.Upload.Upload(ctx, img_path, uploader.UploadParams{Folder: os.ExpandEnv("CLOUDAINARY_STORAGE_FOLDER")})
+	if err != nil {
+		return "", err
+	}
+
+	return uploadResult.SecureURL, nil
+}
+
+func (ctrl ProductController) AddProduct(c *gin.Context) {
+
+	// Get the email of the user to compare with email in the header
+	var input models.AddProductInput
+	err := c.ShouldBind(&input)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": err.Error(), "success": false})
+		return
+	}
+
+	var user models.User
+	if db.DB.Find(&user, "id=?", input.UserID).RecordNotFound() {
+		c.AbortWithStatusJSON(http.StatusNotFound, gin.H{"error": "Something Went Wrong", "success": false})
+		return
+	}
+
+	email := c.Request.Header.Get("email")
+	if email != user.Email {
+		c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "You are not authorized to perform this action", "success": false})
+		c.Abort()
+		return
+	}
+
+	// Extract image from the form
+	r := c.Request
+	r.ParseMultipartForm(10 << 20)
+
+	file, _, err := r.FormFile("image")
```

Comment:
Shouldn't we store multiple images for each product?

Reply:
This option will be discussed later inshallah with khaled